File: internal/billing/secrets.go
package billing

import (
	"context"
	"errors"
	"fmt"
	"strings"

	"github.com/heros-foreal/agentd/internal/providergateway"
)

// secrets.go sources the billing provider's API key and its webhook signing secret from a SECRETS
// MANAGER — never from code, git-tracked config, or telemetry (task 3.3 / design Decision 10).
//
// ## Why this reuses providergateway.Secrets instead of introducing a billing secrets mechanism
//
// The platform already has exactly one answer to "where do credentials come from at call time": the
// `providergateway.Secrets` interface, with an AWS Secrets Manager implementation, a fail-closed
// factory, and a `Describe()` that makes the live source externally readable on /readyz. A second
// mechanism for billing would split that truth in two, and the failure mode is specific and ugly — a
// deployment whose LLM credentials come from a manager while its BILLING credentials quietly come from
// an environment variable, with a health endpoint that is confidently wrong about both.
//
// So billing secrets are ordinary entries in the same source, under reserved logical names. Billing
// gains the manager, the caching, the fail-closed behaviour and the health signal for free, and the
// scrubber that already keeps provider keys out of telemetry covers these by construction.
//
// ## What is deliberately absent
//
// There is no field, method, or struct here that returns a secret for logging, formatting, or
// inclusion in an event. `Describe` returns the SOURCE, never the value. The only way to obtain a
// secret is to ask for it at the moment of use.

// Reserved logical names under which the secrets manager holds the billing credentials. They are
// constants, not literals at call sites, for the same reason metric names are: a name spelled two ways
// is a credential fetched from a place nobody provisioned.
const (
	// SecretBillingAPIKey is the billing provider's server-side API key.
	SecretBillingAPIKey = "billing_provider"
	// SecretBillingWebhookSigning is the webhook signing secret used to verify inbound payloads.
	SecretBillingWebhookSigning = "billing_webhook"
)

// ErrSecretUnavailable is returned when a billing secret cannot be sourced. It FAILS CLOSED: no key
// means no provider call and no webhook verification, never a fallback to an unauthenticated path.
var ErrSecretUnavailable = errors.New("billing: credential unavailable from the secrets manager")

// Secrets supplies the billing provider's credentials at the moment of use.
type Secrets interface {
	// APIKey returns the provider API key.
	APIKey(ctx context.Context) (string, error)
	// WebhookSigningSecret returns the secret inbound webhook signatures are verified against.
	WebhookSigningSecret(ctx context.Context) (string, error)
	// Describe names the SOURCE for the readiness surface — never a secret, never a secret's id.
	Describe() providergateway.SourceInfo
}

// ManagedSecrets adapts a providergateway.Secrets source to the billing capability's two credentials.
type ManagedSecrets struct{ src providergateway.Secrets }

// NewManagedSecrets wraps a secrets source. A nil source is an error rather than a silently
// credential-less service: "the billing integration came up with no secrets source" must be loud at
// construction, not discovered at the first charge.
func NewManagedSecrets(src providergateway.Secrets) (*ManagedSecrets, error) {
	if src == nil {
		return nil, errors.New("billing: a secrets source is required — billing credentials are never read from code or config")
	}
	return &ManagedSecrets{src: src}, nil
}

// APIKey returns the billing provider's API key.
func (m *ManagedSecrets) APIKey(ctx context.Context) (string, error) {
	return m.fetch(ctx, SecretBillingAPIKey)
}

// WebhookSigningSecret returns the webhook signing secret.
func (m *ManagedSecrets) WebhookSigningSecret(ctx context.Context) (string, error) {
	return m.fetch(ctx, SecretBillingWebhookSigning)
}

func (m *ManagedSecrets) fetch(ctx context.Context, name string) (string, error) {
	cred, err := m.src.Credential(ctx, name)
	if err != nil {
		// The error names WHICH credential was missing, never its value or its id.
		return "", fmt.Errorf("%w: %s: %v", ErrSecretUnavailable, name, err)
	}
	key := strings.TrimSpace(cred.APIKey)
	if key == "" {
		return "", fmt.Errorf("%w: %s resolved empty", ErrSecretUnavailable, name)
	}
	return key, nil
}

// Describe names the live secrets source.
func (m *ManagedSecrets) Describe() providergateway.SourceInfo { return m.src.Describe() }

File: web/console/src/lib/idp/secrets.ts
import "server-only";
import { readFile } from "node:fs/promises";
import { join, basename } from "node:path";

/**
 * secrets.ts is the console's `Secrets` seam for IDENTITY credentials (P22 Decision 6, task 5.3).
 *
 * # Why the console needs its own implementation of a seam the platform already has
 *
 * `providergateway.Secrets` is Go, and it is reached at the moment of use — inside the process that
 * holds the credential. The console is a separate process in a separate language, so it cannot import
 * that seam; what it can do, and what `secrets-baseline.md` §1.1 actually asks for, is present the
 * SAME contract with the SAME externally-checkable answer: `HEROS_SECRETS_SOURCE` selects a source,
 * `Describe()` names it, and `/api/health` reports the name so an operator can tell — from the running
 * system, not from a manifest — whether a manager is involved at all.
 *
 * The failure §1.1 was written against is exactly the one to avoid here: a deployment whose LLM and
 * billing credentials come from a manager while the OIDC client secret quietly comes from an
 * environment variable, with the health surface confidently reporting the manager.
 *
 * # The two sources, and why there is no AWS client in this process
 *
 * | kind   | where the value is read from                  | bootstrap secret |
 * |--------|-----------------------------------------------|------------------|
 * | `env`  | a secrets-manager-INJECTED environment variable | none — the injector authenticated |
 * | `file` | a file under `HEROS_SECRETS_DIR`, read at the moment of use | none — the mount was granted by workload identity |
 *
 * `file` is not a lesser option; it is how a managed deployment actually reaches AWS Secrets Manager
 * from a process that should not carry an AWS client: the Secrets Store CSI driver (or a Vault agent
 * file sink, or an air-gapped operator's own mount) projects the manager's value onto a path, the pod
 * is authorised for it AMBIENTLY by its service account, and nothing anywhere holds a long-lived key
 * to reach the manager. That is the property `secrets-baseline.md` §1.1 calls the L1 tiebreak — a
 * manager reached with a long-lived key in an env var has moved the secret, not removed it — and it is
 * preserved here without adding a dependency to a browser-facing process.
 *
 * Reading at the moment of use (rather than at boot) is what makes rotation work: a rotated file is
 * picked up by the next sign-in, with no restart and no cached copy of the old value in memory.
 *
 * # What is deliberately absent
 *
 * Nothing returns a secret for logging, formatting, or inclusion in an event. `describeSecrets()`
 * names the SOURCE — never a value, never a path that contains one, never a secret's id.
 */

/** Reserved logical names. Constants, because a name spelled two ways is a credential nobody provisioned. */
export const SECRET_OIDC_CLIENT_SECRET = "console_idp_client_secret";
export const SECRET_SAML_SP_PRIVATE_KEY = "console_saml_sp_private_key";

/** SECRET_NAMES is every logical name this console reads, so a checklist iterates the real set. */
export const SECRET_NAMES = [SECRET_OIDC_CLIENT_SECRET, SECRET_SAML_SP_PRIVATE_KEY] as const;

/**
 * ErrSecretUnavailable fails CLOSED. No client secret means no code exchange and therefore no session
 * — never a fallback to an unauthenticated token request, and never a weaker mechanism.
 */
export class SecretUnavailableError extends Error {
  constructor(name: string, detail: string) {
    // Names WHICH credential was missing, never its value.
    super(`identity credential unavailable from the secrets source: ${name}: ${detail}`);
    this.name = "SecretUnavailableError";
  }
}

const SOURCE = (process.env.HEROS_SECRETS_SOURCE ?? "env").trim().toLowerCase();
const SECRETS_DIR = process.env.HEROS_SECRETS_DIR ?? "/var/run/secrets/heros";

/**
 * SECRET_PATHS maps a logical name to where this deployment keeps it — CONFIGURATION, not code
 * (task 7.2).
 *
 * The default below is a NAMING CONVENTION, not a mapping table: logical name → same-named env var /
 * same-named file. A deployment whose manager uses different paths overrides the whole map with
 * `CONSOLE_IDP_SECRET_MAP`, so adding a provider is a config injection rather than a new `case` in a
 * switch nobody can change without a release.
 */
function secretLocation(name: string): string {
  const overrides = process.env.CONSOLE_IDP_SECRET_MAP;
  if (overrides) {
    let parsed: unknown;
    try {
      parsed = JSON.parse(overrides);
    } catch {
      throw new Error("CONSOLE_IDP_SECRET_MAP is set but is not valid JSON");
    }
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
      throw new Error("CONSOLE_IDP_SECRET_MAP is set but is not a JSON object of logical name → location");
    }
    const at = (parsed as Record<string, unknown>)[name];
    if (at !== undefined) return String(at);
  }
  return name.toUpperCase();
}

/** identitySecret resolves one identity credential at the moment of use, or fails closed. */
export async function identitySecret(name: string): Promise<string> {
  const location = secretLocation(name);
  if (SOURCE === "file") {
    // `basename` on purpose: a location is a NAME under the secrets directory, never a path that can
    // climb out of it. A mapping document is configuration, and configuration should not be able to
    // read `/etc/shadow` because somebody wrote `../../` in it.
    const path = join(SECRETS_DIR, basename(location));
    try {
      const value = (await readFile(path, "utf8")).trim();
      if (!value) throw new Error("resolved empty");
      return value;
    } catch (err) {
      // The error names the logical name and the source kind, never the path's contents.
      throw new SecretUnavailableError(name, err instanceof Error ? err.message : "unreadable");
    }
  }
  if (SOURCE !== "env") {
    // An unknown source is a refusal, not a silent fall back to `env`. 🔴 no-lazy-defaults: falling
    // back here would make `HEROS_SECRETS_SOURCE=aws-secretz-manager` a typo that silently downgrades
    // a deployment's whole secret posture while /api/health still reports something plausible.
    throw new SecretUnavailableError(name, `unknown HEROS_SECRETS_SOURCE "${SOURCE}"`);
  }
  const value = (process.env[location] ?? "").trim();
  if (!value) throw new SecretUnavailableError(name, "not present in the injected environment");
  return value;
}

/** SecretsSourceInfo mirrors the platform's `/readyz → secrets_source` shape. Never a value. */
export type SecretsSourceInfo = { kind: string; detail: string };

/** describeSecrets names the live source for the health surface. */
export function describeSecrets(): SecretsSourceInfo {
  if (SOURCE === "file") return { kind: "file", detail: SECRETS_DIR };
  return { kind: "env", detail: "secrets-manager-injected environment" };
}
